onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -54,11 +54,61 @@ static common::Status AllocateBufferUsingDeviceAllocatorFromShapeAndType(const T
   return Status::OK();
 }
 
+// null (do nothing) allocator for tensors that wrap external data pointers which are
+// memory mapped or allocated without an ORT allocator; these pointers are instead
+// released with the deleter of the ORT value which holds the ext data tensor
+struct ExtDataNullAllocator : public IAllocator {
+  ExtDataNullAllocator() : IAllocator(OrtMemoryInfo(CPU, OrtAllocatorType::OrtDeviceAllocator)) {}
+  void* Alloc(size_t size) override { ORT_UNUSED_PARAMETER(size); return nullptr; }
+  void Free(void* p) override { ORT_UNUSED_PARAMETER(p); }
+};
+
+// deleter for external data tensors managed by an OrtValue; manages the release of
+// the tensor's data buffer (which points to the external data) and the tensor itself
+struct ExtDataValueDeleter {
+  OrtCallback ext_delete_cb;
+  Tensor* p_tensor;
+  void operator()(void*) noexcept {
+    this->ext_delete_cb.f(this->ext_delete_cb.param);
+    delete this->p_tensor;
+  }
+};
+
+// given a tensor proto with externdal data return an OrtValue with a tensor for
+// that data; the pointers for the tensor data and the tensor itself are owned
+// by the OrtValue's deleter
+static inline common::Status ExtDataTensorProtoToTensor(const Env& env, const std::basic_string<PATH_CHAR_TYPE>& proto_path,
+                                                        const ONNX_NAMESPACE::TensorProto& tensor_proto,
+                                                        Tensor& tensor, OrtCallback& ext_data_deleter)
+{
+  ORT_ENFORCE(utils::HasExternalData(tensor_proto));
+  ORT_ENFORCE(! proto_path.empty());
+
+  void* ext_data_buf = nullptr;
+  size_t ext_data_len = 0;
+  ORT_RETURN_IF_ERROR(utils::GetExtDataFromTensorProto(env, proto_path.c_str(), tensor_proto,
+                                                       ext_data_buf, ext_data_len, ext_data_deleter));
+
+  // NB: creating a do-nothing allocator per tensor is wasteful; can perhaps be
```

Comment:
Doesn't this ctyor of the tensor ensure that the Tensor just owns passed along memory and does nothing to destroy this memory on tensor derstruction - https://github.com/microsoft/onnxruntime/blob/c0dd9be7ba18c19bec5dbc90d3c9e2ec47f24c3d/include/onnxruntime/core/framework/tensor.h#L62 ? In essence, isn't that what you want ?

Reply:
You're right, using this ctor eliminates the need for the custom null deleter.
Switched to this ctor instead.